src/persist-client/src/cli/inspect.rs

```diff
@@ -630,6 +643,204 @@ pub async fn unreferenced_blobs(args: &StateArgs) -> Result<impl serde::Serializ
     Ok(unreferenced_blobs)
 }
 
+/// Roundtrips all blobs through arrow-rs and ensures they match what is currently in S3.
+pub async fn arrow_rs_roundtrip(blob_uri: &str) -> Result<impl Serialize, anyhow::Error> {
+    let cfg = PersistConfig::new_default_configs(&READ_ALL_BUILD_INFO, SYSTEM_TIME.clone());
+
+    let metrics_registry = MetricsRegistry::new();
+    let metrics = Arc::new(Metrics::new(&cfg, &metrics_registry));
+
+    #[derive(Debug, Default, Serialize)]
+    struct BlobReport {
+        num_shards: usize,
+        num_blobs: usize,
+        missing_blobs: usize,
+        num_rows: usize,
+        num_bytes: usize,
+    }
+
+    /// Helper method that lists all shards in a Blob store.
+    async fn list_shards(
+        blob: Arc<dyn Blob + Send + Sync>,
+    ) -> Result<BTreeMap<ShardId, Vec<PartialBatchKey>>, anyhow::Error> {
+        let mut shards = BTreeMap::new();
+        blob.list_keys_and_metadata(&BlobKeyPrefix::All.to_string(), &mut |metadata| {
```

Comment:
Doing this based on blob discovery is clever! Neatly sidesteps a bunch of issues around discovering and trying to read the State of extremely old leaked shards (that certainly exist in prod).

Reply:
Thanks! AFAICT it also avoids any issues with accidentally bumping the state of a persist shard to be registered with a newer version of MZ, truly makes this command read-only